### ISSUE TITLE
How do I get inline diagnostics to work?

### ISSUE THREAD

**Issue by ejulen** [ul/kak-lsp#68]
Hello. I am having issues getting inline diagnostics to work. Even when I use very incorrect syntax for example, I get no visual indication whatsoever that anything is wrong (the errors are listed when I run `:lsp-diagnostics`, though). 

I am using version `4.0.0-rc2`. Here's my `kakrc`:

```
%sh{kak-lsp --kakoune -s $kak_session}
lsp-start
lsp-auto-hover-enable
add-highlighter global number_lines -relative -hlcursor
```

My `kak-lsp.toml` is just the one that comes in the release tarball. What could I be doing wrong?

**Comment by ul** [comment]
Hello!

First thing to check is Kakoune's `*debug*` buffer. It could be that kak-lsp produces malformed command for highlighting diagnostics in your case.

Next thing to check is if diagnostics highlighting command is sent to editor at all. Look at [Troubleshooting](https://github.com/ul/kak-lsp/#troubleshooting) for details how to start kak-lsp with debug logging enabled. As an alternative, to not restart editor, if you have a named Kakoune session you can just kill kak-lsp process and start `kak-lsp -vvv -s <kakoune_session_name> > /tmp/kak-lsp.log` in terminal manually.

Then when working on file which should get inline diagnostics you should get something like

```
Jun 24 07:29:51.625 DEBG To editor `default`: eval -buffer %§/Users/rprakapchuk/kak-lsp/src/controller.rs§ %§set buffer lsp_errors "1459:43.21,43.21|DiagnosticWarning:44.17,44.17|DiagnosticWarning:44.17,44.20|DiagnosticError:43.21,43.21|DiagnosticWarning"§, module: kak_lsp::editor_transport:75
```

in the log file. Keywords to search are `DiagnosticWarning` and `DiagnosticError`.

If everything is fine with previous steps, then it worth checking faces `DiagnosticWarning` and `DiagnosticError`. You can try to override their colors by yourself to be sure that it's not a color issue. By default those faces a set as:

```
set-face global DiagnosticError red
set-face global DiagnosticWarning yellow
```

Another thing which could go wrong is setting highlighter. You can try to set `lsp_errors` highlighter manually with command like 

```
set-option global lsp_errors "%val{timestamp}:1.1,1.3|red"
```

and see if it would highlight first characters in your file. If it wouldn't, then try to execute

```
add-highlighter global/ ranges lsp_errors
```

If this command will enable highlighting then something goes wrong during execution of `%sh{kak-lsp --kakoune -s $kak_session}` output.

If nothing will give a clue of a problem cause, let me know with which language server do you use kak-lsp and I will try to reproduce bug by myself.

**Comment by ejulen** [comment]
First off, thank you for such a detailed response! I can't believe I didn't think of checking the `*debug*` buffer. Something definitely seems to be off, as I find multiple instances of this kind of message in a file with incorrect syntax:

`error running command 'error running command 'eval -buffer %§<file name here>s§ %§set buffer lsp_errors "1:1.2,1.2|DiagnosticError:1.4,1.4|DiagnosticError:1.1,1.0|DiagnosticError"§': 1:1: 'eval' 1:1: 'set' coordinates elements should be >= 1` where `<file name here>` is the file path of the current file.

It's strange, after messing around a bit I sometimes seem to get the inline diagnostics to show up. Still get loads of errors like the one above, though.

Oh, and turns out I had modified the `kak-lsp.toml` file to use [javascript-typescript-langserver](https://github.com/sourcegraph/javascript-typescript-langserver) (which I've used without issues in Neovim) instead of the default `flow-language-server`. On a whim I tried [typescript-language-server](https://github.com/theia-ide/typescript-language-server) just now and my issues seem to have gone away. Maybe the first one I tried is buggy and returns incorrect data to `kak-lsp` sometimes 🤔 

Either way, I think we can consider my issue resolved. If you wish to recreate this issue for yourself, I just had a JS file with the following contents and the aforementioned `javascript-typescript-langserver` as my language server for JS:

```js
f][;
```

This will immediately be marked as incorrect with `typescript-language-server`, but not with `javascript-typescript-langserver` which will instead output errors like the one above.

**Comment by ul** [comment]
This behaviour was a result of bugs in both kak-lsp and javascript-typescript-langserver. kak-lsp was not handling properly cases when character of the end of the range is 0, which should signal that EOL of previous line should be included in range. I partially fixed it in a4be7d8fac57caa1462c93f83f2675745402555c — now it doesn't fail to highlight, but still misbehaves by highlighting one extra character on the next line.

But actual range returned by javascript-typescript-langserver was not exploiting that spec feature, it was just wrong range from line 0 char 0 to line 0 char 0 (even if `f][;` was not on the first line (though even in that case 0.0-0.0 is an invalid range)). I raised an issue for it: https://github.com/sourcegraph/javascript-typescript-langserver/issues/482
